mlir-tensorrt/compiler/include/mlir-tensorrt/Compiler/Client.h

```diff
@@ -145,8 +172,43 @@ class CompilerClient {
   /// used to create the PM.
   llvm::DenseMap<PassManagerKey, std::unique_ptr<CompilationTaskBase>>
       cachedPassManagers;
+
+  static llvm::ManagedStatic<llvm::StringMap<OptionsConstructorFuncT>> registry;
 };
 
+/// Helper to register option types with the client
+template <typename OptionsT, typename TaskT>
+StatusOr<std::unique_ptr<mlir::OptionsContext>>
+optionsCreateFromArgs(const CompilerClient &client,
+                      const std::vector<llvm::StringRef> &args) {
+  // Load available extensions.
+  mlir::MLIRContext *context = client.getContext();
+  mlir::plan::PlanDialect *planDialect =
+      context->getLoadedDialect<mlir::plan::PlanDialect>();
+  compiler::TaskExtensionRegistry extensions =
+      planDialect->extensionConstructors.getExtensionRegistryForTask<TaskT>();
+
+  auto result = std::make_unique<OptionsT>(std::move(extensions));
+
+  std::string err;
+  if (failed(result->parse(args, err))) {
+    std::string line = llvm::join(args, " ");
+    return getInternalErrorStatus(
+        "failed to parse options string {0} due to error: {1}", line, err);
+  }
+
+  // TODO: Figure out whether to add a method in the base class like
+  // "finalizeOptions" or a callback here, or something else if
+  // `inferDeviceOptionsFromHost` is unique to StableHLO.
+  //
+  // Populate device options from host information.
+  Status inferStatus = result->inferDeviceOptionsFromHost();
```

Comment:
Not sure if `inferDeviceOptionsFromHost` is specific to the StableHLO options or not. If it is, I can add a method in the base class to allow child classes to do arbitrary custom things here.

Reply:
The "device" options (e.g. compute capability, max shared mem, etc) and the population of the device options  from local CUDA device should be moved to the common base class.

However, we shouldn't unconditionally invoke `inferDeviceOptionsFromHost()`. That should be a behavior that is moved into the base class constructor. Currently I think we are using a flag `--infer-device-options-from-host`, but we're not doing any validation to check whether the user passed both `--infer-device-options-from-host` as well as `--device-compute-capability=XY`. It may be better to remove `--infer-device-options-from-host` and instead have that be signaled by `--device-compute-capability=0` (which can be the default value).

Reply:
As discussed in the meeting, I can do this in a follow-up PR.